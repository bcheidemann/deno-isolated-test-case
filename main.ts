import { assertRejects } from "@std/assert";
import {
  IsolatedTestContext,
  IsolatedTestContextWithConnection,
} from "./src/IsolatedTestContext.ts";
import { runTestCase } from "./src/runTestCase.ts";
import { spawnIsolatedTestEnvironment } from "./src/spawnIsolatedTestEnvironment.ts";

export type Options = {
  assertFailure: boolean | ((error: unknown) => void | Promise<void>);
};

const ctx = IsolatedTestContextWithConnection.fromEnv();
let wasTestRun = false;
addEventListener("unload", () => {
  if (wasTestRun) {
    return;
  }
  console.error("ERROR: NO TOP-LEVEL TEST DETECTED");
  Deno.exitCode = 1;
});

export function isolatedTestCase(
  name: string,
  fn: (t: IsolatedTestContext) => void | Promise<void>,
  options?: Options,
): void {
  if (ctx) {
    if (ctx.name !== name) {
      return;
    }
    wasTestRun = true;
    runTestCase(ctx, fn);
    return;
  }

  Deno.test({
    name,
    permissions: {
      run: ["deno"],
      net: true,
    },
    fn: async (t) => {
      if (options?.assertFailure) {
        const error = await assertRejects(() =>
          spawnIsolatedTestEnvironment(t)
        );
        if (typeof options.assertFailure === "function") {
          await Promise.resolve(options.assertFailure(error));
        }
      } else {
        await spawnIsolatedTestEnvironment(t);
      }
    },
  });
}
