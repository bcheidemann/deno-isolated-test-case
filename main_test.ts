import { assertEquals, assertInstanceOf } from "@std/assert";
import { isolatedTestCase } from "./main.ts";

if (!Deno.env.get("DENO_ISOLATED_TEST_CASE_CTX")) {
  isolatedTestCase("should fail if test is not run in isolated context", () => {
    assertEquals(1, 1);
  }, {
    assertFailure: true,
  });
}

isolatedTestCase("basic pass", () => {
  assertEquals(1, 1);
});

isolatedTestCase("baisc fail", () => {
  assertEquals(2, 1);
}, { assertFailure: true });

isolatedTestCase("assert error", () => {
  assertEquals(2, 1);
}, {
  assertFailure: (error) => {
    assertInstanceOf(error, Error);
    assertEquals(error.name, "AssertionError");
  },
});
