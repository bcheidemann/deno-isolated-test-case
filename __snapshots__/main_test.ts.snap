export const snapshot = {};

snapshot[`snapshot failure 1`] = `
{
  hello: "world!",
}
`;
